WordPress/gutenberg
# DateTimePicker: Replace react-dates and moment with useLilius and date-fns

## What?

Follows https://github.com/WordPress/gutenberg/pull/41385.

Updates `DatePicker` to use `useLilius` instead of `react-dates` to render its calendar and removes `moment` as a dependency from `@wordpress/components` in favour of `date-fns`.

https://github.com/its-danny/use-lilius
https://github.com/date-fns/date-fns

## Why?

### Problems with `react-dates`

- It contains lots of different components, most of which we do not use. We've had problems with tree shaking in the past.
- It uses `moment` which is quite large and no longer actively developed.
- It requires hacks.
  - We have to use `useEffect` to set an `aria-label` on the day button.
  - We have to do weird stuff to keep focus within the calendar when navigating between months.
  - We have to force re-render by remounting the component using `key` when the date is changed. (This makes the whole thing feel really slow.)
- Limited support for using `@wordpress/components` components. Many of the UI elements can't be swapped out with `@wordpress/components` components. This results in a less consistent UI and more CSS shipped to the browser.

### Considering `react-calendar`

I looked at using `react-calendar` in https://github.com/WordPress/gutenberg/pull/41385. This is a smaller alternative to `react-dates` that uses `date-fns`.

Pros:

- Much fewer bytes.
- No hacks required.
- Uses `date-fns`, allowing us to remove `moment` as a dependency. 

Cons:

- Limited support for using `@wordpress/components` components. Many of the UI elements can't be swapped out with `@wordpress/components` components. This results in a less consistent UI and more CSS shipped to the browser.

### Considering `useLilius`

I then looked at `useLilius` per a suggestion by @stokesman. This is a "headless" library that uses `date-fns`. It offers a single hook that can be used to display a calendar and not any actual UI components.

Pros:

- Barely any bytes.
- No hacks required.
- Uses `date-fns`, allowing us to remove `moment` as a dependency. 
- Headless, so we can use `@wordpress/components` components and Emotion for all of the UI.

Cons:

- Doesn't implement keyboard navigation using the arrow keys.

### Considering `Temporal`

The `Temporal` API is [now stage 3](https://tc39.es/proposal-temporal/docs/), so I considered a completely bespoke implementation using `Temporal`.

Pros:

- Much better data model. `Temporal.PlainDate` is perfectly suited for implementing a calendar as it does not contain time or timezone information, both of which are not relevant in the context of choosing a date from a calendar.
- Future proof. `Temporal` will assumedly be a native browser API in due course.

Cons:

- Slightly more upfront work.
- There's literally a big yellow banner at the top of the proposal saying don't use `Temporal` in production.
- Requires a heavy polyfill. No browser currently implements `Temporal`. What's more,  `Temporal` is much more difficult to tree shake than `date-fns` because all of the functions are implemented as methods on e.g. `Temporal.PlainDate.prototype`. This means the polyfill we ship is quite large.

### Conclusion: `useLilius` + `date-fns`

I settled on using `useLilius` and rolling our own implementation of keyboard navigation. This approach offers the most upfront benefit and is relatively straightforward to implement.

This approach doesn't preclude a later switch to `Temporal`. When the `Temporal` proposal is more developed and implemented by at least one widely used browser (e.g. Chrome), it will be less likely that a user needs to download a heavy polyfill. We should then be able to easily implement our own version of `useLilius` that uses `Temporal.PlainDate`. If you look at [the source of `useLilius`](https://github.com/its-danny/use-lilius/blob/main/src/use-lilius.ts) you'll see that it's quite small.

## How?

- `@wordpress/components`: Remove `moment` and `react-dates` dependencies. Add `date-fns` and `use-lilius`.
- `DateTimePicker`: Use `VStack` so that we can remove all CSS.
- `TimePicker`: Swap `moment` functions for `date-fns` functions.
- `DatePicker`
  - Use `useLilius` instead of `react-dates`.
  - Use `@wordpress/components` and Emotion instead of CSS.
  - Use `dateI18n` so that dates are correctly translated. 
  - Use `wp.date.getSettings()` so that dates appear in the selected format.
  - Add custom implementation of keyboard navigation.

## Testing Instructions

1. Create or edit a post.
2. Open the _Publish_ popover in the document settings sidebar.
3. Have at it. Things should work mostly the same as they did before. Be sure to test keyboard navigation and using screen reader.

## Screenshots or screencast 

Before:

https://user-images.githubusercontent.com/612155/182987676-dbd31296-eff6-47ad-9d9e-799e137b2b68.mp4

After:

https://user-images.githubusercontent.com/612155/182987811-486a1e81-824f-4581-b8f6-f832b8f82c12.mp4

I don't know if it comes across on video but it feels much snappier to use because we're no longer remounting the component 🙂
